rfd/0201-native-auto-approval.md

```diff
@@ -0,0 +1,364 @@
+---
+authors: Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0201 - Native Auto Approval
+
+## Required Approvers
+* Engineering: @r0mant && @fheinecke
+
+## What
+This document describes how Teleport will support native automatic approvals for
+access requests.
+
+## Why
+Currently, Teleport supports automatic approvals of access requests, but support
+is limited. Automatic approvals requires a separate plugin, and only a subset of
+the plugins support automatic approvals. Users requesting this feature do not
+wish to maintain a separate piece of software, or wish to integrate with a
+separate service to utilize this feature. In order to support a wider range of
+use cases, Teleport should support automatic approvals natively.
+
+Automatic approvals enables teams to enforce zero standing privileges, while
+allowing users to get access to their pre approved resources for a limited
+period of time. There are alternative ways to enforce zero standing privileges,
+but a big part of this feature request is to provide a frictionless user
+experience. For example, access lists can be used to achieve similar behavior,
+but a limitation with access lists is that the user must logout and login again
+to renew their certificates to obtain the new roles.
+
+This feature has also been requested for internal use at Teleport. The Tools
+team enforces gated access to certain pipeline activity. The team would like the
+dev environment to closely mimic the production environment and would like to
+enforce the same gated access, but it would reduce a lot of friction if access
+requests can be automatically approved in the dev environment.
+
+## Goals
```

Comment:
This sounds great! Couple of questions:
* When will Terraform support for the changes be added? This is how we deploy all internal Teleport resources currently, and will be needed to adopt this internally.
* I see "user" used throughout this document. Will this support tbot roles/service users, or just "real" users? For our internal use case, the requests will come from a tbot service user.

Reply:
- The Teleport Terraform provider already supports access monitoring rules, but the tf resource schema will need to be updated with the new `automatic_approval` field. I'll add a note about Terraform support, and add an additional step in the **Implementation Plan** section.
- I'm not too familiar with the capabilities of a bot user, but I would assume they would have the ability to make an access request just like a regular user, assuming that they have the correct role permissions. I'll investigate a bit more on this topic and add a section.

Reply:
I might be wrong about tbot use case. It looks like Teleport prohibits access requests when using an impersonated identity. I'll see if there is a workaround.
https://github.com/gravitational/teleport/blob/dac5a303968683a0e2e6c5a554a6a807e99272e6/lib/auth/auth_with_roles.go#L3204-L3218

Reply:
I'll talk with @r0mant about this today and get back to you.

> this method might reduce some of the security benefits of using machine id

Do you know what security benefits specifically would be impacted?

Reply:
From the [RFD](https://github.com/gravitational/teleport/blob/master/rfd/0064-bot-for-cert-renewals.md#security), tbot leverages impersonation to split the responsibilities for renewal and access. This allows users to mange their renewable and non-renewable certificates separately, with different levels of security/access.

Not sure if these concerns apply to our use case.

Reply:
@bernardjkim I've got this on my TODO list to test. For now let's assume that it will work.

I missed your previous question:
> Would this approach support your use case?

I think so. The `tbot` user will assume a role that can both create new roles and file access requests for them. I think once the tbot user assume the role, it should act the same as a "normal" Teleport user.

Reply:
Fwiw, right now, Machine ID really doesn't support Access Requests. Machine ID bots are permitted to review access requests, but, they can't impersonate them. I think I'm missing some context around this RFD, what would be the use-case for a Bot using Access Requests/automatic approval?

Reply:
@fheinecke @bernardjkim @strideynet There may be some misunderstanding - for our internal use-case (release approval service), we don't need Machine ID to assume access requests. We just need to be able to create a role, submit an access request for it and get it auto-approved if it matches a certain criteria. That's it, no need to assume the approved request or anything, the service will just check the access request status to see if the release is approved or not.

Reply:
I'm going to defer further investigation on this topic. If access requests are compatible with Machine ID, automatic approvals will also be compatible.

From what I understand, bot users cannot assume an access request and it cannot create an access request.

- Teleport prohibits impersonated users from assuming new roles. https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L3204
- Teleport access request validation logic only looks at "statically assigned" roles to check if the user is permitted to create an access request. https://github.com/gravitational/teleport/blob/master/lib/services/access_request.go#L2017

Reply:
> From what I understand, bot users with impersonated roles cannot assume an access request and it cannot create an access request.

Correct - but I don't understand why a Bot needs to create or assume an access request.

Reply:
It might be better to continue discussing this topic in https://github.com/gravitational/teleport.e/pull/5533.